contracts/integrations/nevermined/BalanceTrackerSubscription.sol

```diff
@@ -0,0 +1,216 @@
+        subscriptionTokenId = _subscriptionTokenId;
+    }
+
+    // Check and record delivery rate
+    function checkAndRecordDeliveryRate(
+        address mech,
+        address requester,
+        uint256,
+        bytes memory
+    ) external payable {
+        // Check for marketplace access
+        if (msg.sender != mechMarketplace) {
+            revert ManagerOnly(msg.sender, mechMarketplace);
+        }
+
+        // Get mech max delivery rate
+        uint256 maxDeliveryRate = IMech(mech).maxDeliveryRate();
+
+        // Check that there is no incoming deposit
+        if (msg.value > 0) {
+            revert NoDepositAllowed(msg.value);
+        }
+
+        // Get requester credit balance
+        uint256 balance = mapRequesterBalances[requester];
+        // Get requester actual subscription balance
+        uint256 subscriptionBalance = IERC1155(subscriptionNFT).balanceOf(requester, subscriptionTokenId);
+
+        // Adjust requester balance with maxDeliveryRate credits
+        balance += maxDeliveryRate;
+
+        // Check the request delivery rate for a fixed price
+        if (subscriptionBalance < balance) {
+            revert InsufficientBalance(subscriptionBalance, balance);
+        }
+
+        // Adjust requester balance
+        mapRequesterBalances[requester] = balance;
+    }
+
+    /// @dev Finalizes mech delivery rate based on requested and actual ones.
+    /// @param mech Delivery mech address.
+    /// @param requester Requester address.
+    /// @param requestId Request Id.
+    /// @param maxDeliveryRate Requested max delivery rate.
+    function finalizeDeliveryRate(address mech, address requester, uint256 requestId, uint256 maxDeliveryRate) external {
+        // Reentrancy guard
+        if (_locked > 1) {
+            revert ReentrancyGuard();
+        }
+        _locked = 2;
+
+        // Check for marketplace access
+        if (msg.sender != mechMarketplace) {
+            revert ManagerOnly(msg.sender, mechMarketplace);
+        }
+
+        // Get actual delivery rate
+        uint256 actualDeliveryRate = IMech(mech).getFinalizedDeliveryRate(requestId);
+
+        // Check for zero value
+        if (actualDeliveryRate == 0) {
+            revert ZeroValue();
+        }
+
+        uint256 rateDiff;
+        if (maxDeliveryRate > actualDeliveryRate) {
+            // Return back requester overpayment credit
+            rateDiff = maxDeliveryRate - actualDeliveryRate;
+
+            // Get requester balance
+            uint256 balance = mapRequesterBalances[requester];
+
+            // This must never happen as max delivery rate is always bigger or equal to the actual delivery rate
+            if (rateDiff > balance) {
+                revert Overflow(rateDiff, balance);
+            }
+
+            // Adjust requester balance
+            balance -= rateDiff;
+            mapRequesterBalances[requester] = balance;
+        } else {
+            actualDeliveryRate = maxDeliveryRate;
+        }
```

Comment:
Here we can finally exactly account for actual delivery rate and reduce credits from requester, if delivery rate cost less.